draft-ietf-quic-recovery.md

```diff
@@ -1581,6 +1588,10 @@ Invoked when DetectAndRemoveLostPackets deems packets lost.
 
 ~~~
    InPersistentCongestion(lost_packets):
```

Comment:
IIUC, this function is expected to check for losses across all packet number spaces. Assuming that this understanding is correct, I think we should not pass `lost_packets` as an argument (to this function or to the `AreAllPacketsLost` that is being called from this function), because `lost_packets` is a per-PN value that is obtained by calling `DetectAndRemoveLostPackets(pn_space)`.

And I think that this mitigates the concern raised in #3831.

Reply:
Agreed, it needs to be across PN spaces.  Done.

Reply:
I disagree. While strictly speaking this change is not wrong, it makes things more complex than necessary: An ACK can only ever establish loss in its own packet number space, never across packet number spaces. So it's totally fine to pass `lost_packets` to `InPersistentCongestion`.

Reply:
My point is that passing `loss_packets` in the pseudo-code gives the readers false impression that persistent congestion is something to be observed per-PN-space.

An endpoint can pass arguments whatever necessary (or beneficial), but that does not mean that those arguments should appear in pseudo-code.

Reply:
The comment now says:
> Determine if all packets in the time period before the largest newly lost packet, including the edges and across all packet number spaces, are marked lost.

My point is that this is unnecessarily complicated, as only packets within a single packet number space, namely the packet number space the ACK was received in, can be newly lost.

Reply:
@marten-seemann I think you are suggesting of writing down in the spec an algorithm that can be used to detect persistent congestion, rather than just having something like `AreAllPacketsLost`. I'm not generally opposed to the idea, though I think we need to first seek what kind of implementation that we want to write down (assuming that we'd be writing down one). I think that deserves a separate issue.

Reply:
That's a separate issue, see #3937.

Independent of that, I still think that 3ddca85d4b885e053a7312f02af4dfb49bb9d3ea doesn't clarifying things, and would prefer to revert it.

Reply:
I think your argument is based on the assumption that `lost_packets` is useful, which I disagree. Please see https://github.com/quicwg/base-drafts/pull/3889#discussion_r458524100.

Reply:
After talking with @kazuho OOB, I managed to work out what our disagreement is about. Consider the following example:

If there’s an ACK that declares loss in 1-RTT space for packets sent between T=0 and T=5 (which, taken for itself would establish persistent congestion), but there was a Handshake packet acknowledged (in a separate ACK frame) sent at T=3, @kazuho wouldn't declare persistent congestion. I very much agree with that logic, and note that this would constitute a fine solution for #3831.

Pseudo-code-wise, I think we'll have to get a bit fancy here, by introducing two new global variables:
* `time_of_first_rtt_measurement`, the timestamp we were first able to establish an RTT measurement
* `largest_acked_packet_send_time`, the timestamp when the largest acknowledged packet was sent, across all packet number spaces (note that this value is only updated **after** calling `InPersistentCongestion`).

When checking for persistent congestion in `InPersistentCongestion(lost_packets)`, you'd only consider packets sent after `max(time_of_first_rtt_measurement, largest_acked_packet_send_time)`. How exactly "considering" here looks like will depend on the outcome of #3937.

Reply:
@marten-seemann It is impossible to rely on `largest_acked_packet_send_time`, because an ACK on another PN space might have arrived later than the persistent congestion period to be declared. See the example below.

![IMG_3331](https://user-images.githubusercontent.com/41567/88139839-81b4ef00-cc2b-11ea-8c19-90abd434ed91.jpeg)

As stated in https://github.com/quicwg/base-drafts/pull/3889#discussion_r458524100, I do not think it is possible can use `loss_packets`, unless you implement some lossy behavior to bound state.

Reply:
@kazuho That makes sense to me. The algorithm I envisioned doesn't work. Not sure how one would correctly implement a `InPersistentCongestion` that yields correct results across packet number spaces then...

Reply:
@marten-seemann IIUC, one trivial way of implementing `InPersistentCongestion` is as follows:
* Retain the entries in sentmap (the list of packets being sent) for at least 3 PTO.
* When receiving an acknowledgement, mark a hole in the sentmap for the packet being acked.
* When a loss is detected in a packet number space, check if that loss spans across more than 3 PTO, by traversing the sentmap. Then, if did, check the sentmap of other packet number spaces (by traversing through those sentmaps) to see if any ACK-eliciting packet sent in that period has been acked.

This approach might sound too trivial, as it is O(N) where N is the number of packets being lost (or in case there are multiple PN spaces in action, the number of inflight packets in other PN spaces). But in practice, I think this approach is sufficient in terms of performance, because N would be like 3 for the current packet number space, and because we would not be sending that many packets on Initial and Handshake packet number spaces.

Reply:
> When receiving an acknowledgement, mark a hole in the sentmap for the packet being acked.

I don't think that's trivial. Creating a "hole" is extra state that has to be cleaned up at some point.

Reply:
Moving the discussion to #3939, starting from https://github.com/quicwg/base-drafts/issues/3939#issuecomment-662808337. @marten-seemann I will respond there.